### PR TITLE
apply system params on next block

### PR DESCRIPTION
**Pull request by kroggen**
If enough votes to change a system param value is reached, save the new value on a temporary place and only update the param value when the block is finished

In this way the new value will only be active for the next block

Also keep the values from params on memory instead of loading params from state on every block

### CODE REVIEW

**Comment by kroggen** [comment]
It would be useful to have some tests for this

**Comment by gokch** [comment]
I add mutex about systemParams because it occur panic when access map in same time.

**Comment by gokch** [comment]
it need additional sync test before merge.

File: contract/system/param.go [hunk 1]
```diff
@@ -86,51 +134,45 @@ func updateParam(s dataSetter, id string, value *big.Int) (error) {
...
+	if val := systemParams.getParam(proposalID); val != nil {
 		return val
 	}
-	return DefaultParams[proposalID]
+	return DefaultParams.getParam(proposalID)
 }
 
 // these 4 functions are reading the param value for the current block
-
 func GetStakingMinimum() *big.Int {
```
Comment (kroggen) [inline contract/system/param.go:182]: This comment is for the next 4 function, so it should have an empty line

File: contract/system/param.go [hunk 2]
```diff
@@ -148,7 +190,6 @@ func GetBpCount() int {
 }
 
 // these functions are reading the param value directly from the state
-
 func GetNamePriceFromState(scs *state.ContractState) *big.Int {
```
Comment (kroggen) [inline contract/system/param.go:199]: Same

File: contract/system/param.go (hunk 1 above)
Comment (gokch) [reply contract/system/param.go:179]: Is these our official convention? all code can be changed to using these convension.

Comment (gokch) [reply contract/system/param.go:179]: personally, i prefer these convension to division line
image

Comment (kroggen) [reply contract/system/param.go:179]: loogs good

**Comment by kslee8282** [comment]
Start sync test (testnet)
Oct 13 05:00:30.592 INF ../go/aergo/cmd/aergosvr/aergosvr.go:100 > AERGO SVR STARTED branch=topic/system-params module=asvr revision=a282031c

**Comment by kslee8282** [comment]
Start sync test (mainnet)
Oct 13 05:05:40.440 INF ../go/aergo/cmd/aergosvr/aergosvr.go:100 > AERGO SVR STARTED branch=topic/system-params module=asvr revision=a282031c

**Comment by kslee8282** [comment]
Sync completed

**Comment by kroggen** [comment]
I am thinking in renaming these functions:

* `getNextParam` -> `getNextBlockParam`
* `setNextParam` -> `setNextBlockParam`
* `delNextParam` -> `delNextBlockParam`

It does not require a new sync test

@hayarobi @gokch  What do you think?

File: contract/system/param.go [hunk 3]
```diff
@@ -26,7 +70,10 @@ const (
 )
 
 var (
-	systemParams parameters
+	systemParams *parameters = &parameters{
+		mutex:  sync.RWMutex{},
+		params: map[string]*big.Int{},
+	}
```
Comment (kroggen) [inline contract/system/param.go:76]: I suspect that the above can be just `systemParams *parameters`
It is loaded on `loadParams()`

**Comment by gokch** [comment]
> I am thinking in renaming these functions:
> 
> * `getNextParam` -> `getNextBlockParam`
> * `setNextParam` -> `setNextBlockParam`
> * `delNextParam` -> `delNextBlockParam`
> 
> It does not require a new sync test
> 
> @hayarobi @gokch What do you think?

i agree.

**Comment by kroggen** [comment]
Can this be merged into develop?

**Review by gokch — APPROVED** [review]

**Pull request merged**